src/nice-path-decomposition/nice-path-decomposition.cpp

```diff
@@ -0,0 +1,51 @@
+#include "nice-path-decomposition.h"
+#include "../../tdlib/src/iter.hpp"
+
+std::vector<NicePathDecomposition::Bag> NicePathDecomposition::get_nice_bags() {
+    return _nice_bags;
+}
+
+void NicePathDecomposition::set_g(Graph &g) {
+    _g = g;
+}
+
+NicePathDecomposition::NicePathDecomposition(std::vector<std::vector<vertex_t>> &bags, Graph &g) : PathDecomposition(
+        bags, g) {
+    nice_transform();
+}
+
+
+NicePathDecomposition::NicePathDecomposition(PathDecomposition pathDecomposition) : PathDecomposition(
+        pathDecomposition) {
+    nice_transform();
+}
+
+void NicePathDecomposition::nice_transform() {
+    std::vector<vertex_t> prev_bag = {};
+    std::set<vertex_t> prev_set = {};
+    for (const std::vector<vertex_t>& bag : _bags) {
+        std::set<vertex_t> vertices;
+        for (vertex_t v : bag) vertices.insert(v);
+        for (int i = 0; i < prev_bag.size(); i++) {
```

Comment:
This is a for loop over a changing vector that manipulates i inside the loop. This makes code harder to read and more bug-prone than necessary. Here you compute the difference between the set of nodes in the current bag and in the next one. There is a function in the STD algorithm library that does that.